File: java/omnia/data/cache/SimpleLongCacher.java
package omnia.data.cache;

import static java.util.Objects.requireNonNull;

import java.util.function.LongSupplier;

/**
 * A {@link CachedLong} implementation that uses the given {@link LongSupplier} to provide the
 * value to be cached. This class is not thread-safe.
 *
 * The given {@link LongSupplier#getAsLong()} method is never invoked if the value is cached.
 */
final class SimpleLongCacher implements CachedLong {
  private final LongSupplier supplier;
  private boolean isValid;
  private long value;

  SimpleLongCacher(LongSupplier supplier) {
    this.supplier = requireNonNull(supplier);
  }

  @Override
  public long value() {
    if (!isValid) {
      value = supplier.getAsLong();
      isValid = true;
    }
    return value;
  }

  @Override
  public void invalidate() {
    isValid = false;
  }
}

File: src/omnia/data/cache/SimpleLongCacher.kt
package omnia.data.cache

import java.util.function.LongSupplier

/**
 * A [CachedLong] implementation that uses the given [LongSupplier] to provide the
 * value to be cached. This class is not thread-safe.
 *
 * The given [LongSupplier.getAsLong] method is never invoked if the value is cached.
 */
internal class SimpleLongCacher(private val supplier: LongSupplier) : CachedLong {

  private var isValid = false
  private var value: Long = 0
  override fun value(): Long {
    if (!isValid) {
      value = supplier.asLong
      isValid = true
    }
    return value
  }

  override fun invalidate() {
    isValid = false
  }
}